parsing/parser.mly

```diff
@@ -785,37 +834,65 @@ open_statement:
   | OPEN override_flag mod_longident post_item_attributes
       { Opn.mk (mkrhs $3 3) ~override:$2 ~attrs:$4 ~loc:(symbol_rloc()) }
 ;
-module_declaration:
+sig_include_statement:
+    INCLUDE module_type post_item_attributes %prec below_WITH
+      { Incl.mk $2 ~attrs:$3 ~loc:(symbol_rloc()) }
+;
+module_declaration_body:
     COLON module_type
       { $2 }
-  | LPAREN UIDENT COLON module_type RPAREN module_declaration
+  | LPAREN UIDENT COLON module_type RPAREN module_declaration_body
       { mkmty(Pmty_functor(mkrhs $2 2, Some $4, $6)) }
-  | LPAREN RPAREN module_declaration
+  | LPAREN RPAREN module_declaration_body
       { mkmty(Pmty_functor(mkrhs "*" 1, None, $3)) }
 ;
-module_rec_declarations:
-    module_rec_declaration                              { [$1] }
-  | module_rec_declarations AND module_rec_declaration  { $3 :: $1 }
+module_declaration:
+    MODULE UIDENT module_declaration_body post_item_attributes
+      { Md.mk (mkrhs $2 2) $3 ~attrs:$4 ~loc:(symbol_rloc()) }
+;
+module_alias:
+    MODULE UIDENT EQUAL mod_longident post_item_attributes
+      { Md.mk (mkrhs $2 2)
+          (Mty.alias ~loc:(rhs_loc 4) (mkrhs $4 4))
+          ~attrs:$5 ~loc:(symbol_rloc()) }
 ;
-module_rec_declaration:
-    UIDENT COLON module_type post_item_attributes
-    { Md.mk (mkrhs $1 1) $3 ~attrs:$4 ~loc:(symbol_rloc()) }
+rec_module_declarations:
+    rec_module_declaration                          { [$1] }
+  | rec_module_declarations and_module_declaration  { $2 :: $1 }
+;
+rec_module_declaration:
+    MODULE REC UIDENT COLON module_type post_item_attributes
+      { Md.mk (mkrhs $3 3) $5 ~attrs:$6 ~loc:(symbol_rloc()) }
+;
+and_module_declaration:
+    AND UIDENT COLON module_type post_item_attributes
+      { Md.mk (mkrhs $2 2) $4 ~attrs:$5 ~loc:(symbol_rloc()) }
+;
+module_type_declaration_body:
+    /* empty */               { None }
+  | EQUAL module_type         { Some $2 }
+;
+module_type_declaration:
+    MODULE TYPE ident module_type_declaration_body post_item_attributes
+      { Mtd.mk (mkrhs $3 3) ?typ:$4 ~loc:(symbol_rloc()) ~attrs:$5 }
 ;
-
 /* Class expressions */
 
 class_declarations:
-    class_declarations AND class_declaration    { $3 :: $1 }
-  | class_declaration                           { [$1] }
+    class_declaration                           { [$1] }
+  | class_declarations and_class_declaration    { $2 :: $1 }
 ;
 class_declaration:
-    virtual_flag class_type_parameters LIDENT class_fun_binding
+    CLASS virtual_flag class_type_parameters LIDENT class_fun_binding
     post_item_attributes
-      {
-       Ci.mk (mkrhs $3 3) $4
-         ~virt:$1 ~params:$2
-         ~attrs:$5 ~loc:(symbol_rloc ())
-      }
+      { Ci.mk (mkrhs $4 4) $5 ~virt:$2 ~params:$3
+         ~attrs:$6 ~loc:(symbol_rloc ()) }
+;
+and_class_declaration:
+    AND virtual_flag class_type_parameters LIDENT class_fun_binding
```

Comment:
Doing the relocation after the fact complicates reading the parser. Whilst the patch does introduce some redundancy between `foo` and `and_foo`, it also means that all information related to a particular `foo` is gathered at a single point in the parser. If you add information separately later, then you must trace the path that the `foo` takes through the parser in order to find all the information attached to it.

The other issue is that -- unlike functions in a normal language -- abstractions in a LALR grammar change the semantics of the grammar. By introducing additional reduce actions you can introduce additional conflicts. This means that removing redundancy by introducing new rules actually increases the difficulty in understanding the parser. In particular, it makes it harder to know whether a particular change might introduce a conflict or not.

So, overall I would rather leave these small redundancies in the parser because I think the result is actually easier to read and reason about.